src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -1846,16 +1846,19 @@ void c_typecheck_baset::typecheck_expr_side_effect(side_effect_exprt &expr)
 
     if(type0.id() == ID_c_enum_tag)
     {
-      if(follow_tag(to_c_enum_tag_type(type0)).is_incomplete())
+      const c_enum_typet &enum_type = follow_tag(to_c_enum_tag_type(type0));
+      if(enum_type.is_incomplete())
       {
         error().source_location = expr.source_location();
         error() << "operator `" << statement
                 << "' given incomplete type `"
                 << to_string(type0) << "'" << eom;
         throw 0;
       }
-      else
-        expr.type()=type0;
+
+      // increment/decrement on underlying type
+      expr.op0() = typecast_exprt(expr.op0(), enum_type.subtype());
+      expr.type() = enum_type.subtype();
```

Comment:
I would make that an implicit typecast.

Reply:
I have added `implicit_typecast_arithmetic(expr);` - is that what you expected?

Reply:
No, that's arithmetic promotion; I am not actually sure whether these are meant to happen to the underlying type of an enum.
What I meant is the `ID_C_implicit_typecast` annotation, which is used to distinguish an explicit cast from one added by the frontend.

Reply:
I can't seem to find `ID_C_implicit_typecast` or indeed any other annotation that the `implicit_typecast` code would be adding?!